src/storage/mvcc/txn.rs

```diff
@@ -132,6 +141,169 @@ impl<S: Snapshot> MvccTxn<S> {
         Ok(self.reader.get_write(&key, ts)?.is_some())
     }
 
+    pub fn acquire_pessimistic_lock(
+        &mut self,
+        key: Key,
+        primary: &[u8],
+        for_update_ts: u64,
+        should_not_exist: bool,
+        options: &Options,
+    ) -> Result<()> {
+        if let Some(lock) = self.reader.load_lock(&key)? {
+            if lock.ts != self.start_ts {
+                return Err(Error::KeyIsLocked {
+                    key: key.into_raw()?,
+                    primary: lock.primary,
+                    ts: lock.ts,
+                    ttl: lock.ttl,
+                });
+            }
+            if lock.lock_type != LockType::Pessimistic {
+                return Err(Error::LockTypeNotMatch {
+                    start_ts: self.start_ts,
+                    key: key.into_raw()?,
+                    pessimistic: false,
+                });
+            }
+            MVCC_DUPLICATE_CMD_COUNTER_VEC.pessimistic_lock.inc();
+            return Ok(());
+        }
+
+        if let Some((commit_ts, write)) = self.reader.seek_write(&key, u64::max_value())? {
+            // The isolation level of pessimistic transactions is RC. `for_update_ts` is
+            // the commit_ts of the data this transaction read. If exists a commit version
+            // whose commit timestamp is larger than current `for_update_ts`, the
+            // transaction should retry to get the latest data.
+            if commit_ts > for_update_ts {
+                MVCC_CONFLICT_COUNTER.pessimistic_lock_conflict.inc();
+                return Err(Error::WriteConflict {
+                    start_ts: self.start_ts,
+                    conflict_start_ts: write.start_ts,
+                    conflict_commit_ts: commit_ts,
+                    key: key.into_raw()?,
+                    primary: primary.to_vec(),
+                });
+            }
+
+            // Handle rollback.
+            // If the start timestamp of write is equal to transaction's start timestamp
+            // as well as commit timestamp, the lock is already rollbacked.
+            if write.start_ts == self.start_ts && commit_ts == self.start_ts {
```

Comment:
Could we just check `write.start_ts=self.start_ts`? And If yes, just return this transaction is commited or rollbacked？

Reply:
Yes. It should check `committed` here.